### PR TITLE
Fixes #9752 - Pass Docker errors to controller

### PR DESCRIPTION
**Pull request by dLobatog**

### CODE REVIEW

**Comment by dLobatog** [comment]
[test]

File: app/models/service/containers.rb [hunk 1]
```diff
@@ -38,20 +38,21 @@ def create_container_object(wizard_state)
         container.send(:"#{taxonomy}=", wizard_state.preliminary.send(:"#{taxonomy}"))
       end
 
-      unless container.valid?
-        @errors = container.errors
-        fail ActiveRecord::Rollback
-      end
+      pull_image(container)
+      start_container(container)
+      errors << container.errors unless container.valid?
```
Comment (ohadlevy) [inline app/models/service/containers.rb]: wont this add a nested array? e.g.
```ruby
irb(main):001:0> [1,2] << [3,4]
=> [1, 2, [3, 4]]
```

File: app/models/service/containers.rb [hunk 2]
```diff
@@ -38,20 +38,21 @@ def create_container_object(wizard_state)
...
       container.name = container.in_fog.name[1..-1] unless container.name.present?
 
       container
     end
 
     def pull_image(container)
-      container.compute_resource.create_image(:fromImage => container.repository_pull_url)
+      success = container.compute_resource.
+        create_image(:fromImage => container.repository_pull_url)
+      errors << container.compute_resource.errors[:base] unless success
```
Comment (ohadlevy) [inline app/models/service/containers.rb]: ditto, use array + array instead?

File: test/units/containers_service_test.rb [hunk 3]
```diff
@@ -24,4 +24,15 @@ class ContainersServiceTest < ActiveSupport::TestCase
...
   end
+
+  test 'passes errors from compute resource' do
+    # Since the compute resource will be unreachable, this test will always
+    # fail at the 'pull_image' step
+    containers_service = Service::Containers.new
+    assert_raises(ActiveRecord::Rollback) do
+      containers_service.create_container_object(@state)
+    end
+    assert containers_service.errors.present?
```
Comment (ohadlevy) [inline test/units/containers_service_test.rb]: would you mind adding the test for the errors themselves?

**Comment by dLobatog** [comment]
Merged as ef3feba424601cb70a7f25fd8457d885e8a023b2, thanks @dLobatog!

File: test/units/containers_service_test.rb (hunk 3 above)
Comment (dLobatog) [inline test/units/containers_service_test.rb]: By that you mean writing a test for each of the possible errors in the ticket (docker unreachable, wrong permissions, image not found, etc...)?

Comment (ohadlevy) [inline test/units/containers_service_test.rb]: well, the array bug mentioned earlier would have been found if you had at least one test discussing the value of the errors array. i don't think we care too much about the actual errors (as they are generated by docker itself) I would only care about errors we raise internally and probably generic handling of errors coming from external sources?

File: app/models/service/containers.rb (hunk 2 above)
Comment (dLobatog) [inline app/models/service/containers.rb]: also in `start_container`

File: test/units/containers_service_test.rb (hunk 3 above)
Comment (dLobatog) [inline test/units/containers_service_test.rb]: https://github.com/theforeman/foreman-docker/pull/168

**Pull request closed without merging**